From 5fae6beb45dfa3459112fa59ca53f3962e3ef5a3 Mon Sep 17 00:00:00 2001
From: Adithya R <gh0strider.2k18.reborn@gmail.com>
Date: Sat, 17 Feb 2024 07:26:49 +0530
Subject: [PATCH 2/3] audio: Check MODIFY_AUDIO_ROUTING permission for per-app
 volume

We need to use this from SystemUI which runs under app uid.
Instead check for MODIFY_AUDIO_ROUTING permission which is granted only
to select system apps (including SystemUI) and also matches the API
@RequiresPermission annotation in framework.

Change-Id: Ib78565c15cbdc8619bf94240a2329e97fd40c6a9
---
 services/audioflinger/AudioFlinger.cpp | 16 ++++++++++++++++
 1 file changed, 16 insertions(+)

diff --git a/services/audioflinger/AudioFlinger.cpp b/services/audioflinger/AudioFlinger.cpp
index 96537fe53c..0da75220ab 100644
--- a/services/audioflinger/AudioFlinger.cpp
+++ b/services/audioflinger/AudioFlinger.cpp
@@ -5322,6 +5322,22 @@ status_t AudioFlinger::onTransactWrapper(TransactionCode code,
             break;
     }
 
+    // make sure the following transactions require MODIFY_AUDIO_ROUTING permission
+    switch (code) {
+        case TransactionCode::SET_APP_VOLUME:
+        case TransactionCode::SET_APP_MUTE: {
+            if (!modifyAudioRoutingAllowed()) {
+                ALOGW("%s: transaction %d received from PID %d UID %d does not have "
+                      "MODIFY_AUDIO_ROUTING permission",
+                      __func__, code, IPCThreadState::self()->getCallingPid(),
+                      IPCThreadState::self()->getCallingUid());
+                return INVALID_OPERATION;
+            }
+        } break;
+        default:
+            break;
+    }
+
     const std::string methodName = getIAudioFlingerStatistics().getMethodForCode(code);
     mediautils::TimeCheck check(
             std::string("IAudioFlinger::").append(methodName),
-- 
2.51.1

